### ISSUE TITLE
`bundle outdated` and `bundle update` can't find some updates when block-syntax source defs exist

### ISSUE THREAD

**Issue by tjgrathwell** [bundler/bundler#3368]
Using this Gemfile:
```
source 'https://rubygems.org'

gem 'httparty', '0.13.1'
source 'https://rails-assets.org' do
  gem 'rails-assets-lodash', '2.4.1'
end
```

which produces this Gemfile.lock on a Bundle,
```
GEM
  remote: https://rubygems.org/
  remote: https://rails-assets.org/
  specs:
    httparty (0.13.1)
      json (~> 1.8)
      multi_xml (>= 0.5.2)
    json (1.8.2)
    multi_xml (0.5.5)
    rails-assets-lodash (2.4.1)

PLATFORMS
  ruby

DEPENDENCIES
  httparty (= 0.13.1)
  rails-assets-lodash (2.4.1)!
```

The output of `bundle outdated` is like this:
```
Fetching gem metadata from https://rails-assets.org/...
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...

Outdated gems included in the bundle:
  * rails-assets-lodash (3.0.1 > 2.4.1) Gemfile specifies "= 2.4.1"
```

Compare to this Gemfile:
```
source 'https://rubygems.org'
source 'https://rails-assets.org'

gem 'httparty', '0.13.1'
gem 'rails-assets-lodash', '2.4.1'
```

Which produces this `bundle outdated`:
```
Fetching gem metadata from https://rails-assets.org/...
Fetching gem metadata from https://rubygems.org/.......
Resolving dependencies...

Outdated gems included in the bundle:
  * httparty (0.13.3 > 0.13.1) Gemfile specifies "= 0.13.1"
  * rails-assets-lodash (3.0.1 > 2.4.1) Gemfile specifies "= 2.4.1"
```

Note that if you have a local  `httparty` gem installed that is greater than 0.13.1, it WILL show up in the `bundle outdated` output. So the best way to reproduce is to use a fresh ruby/gemset.

**Comment by tjgrathwell** [comment]
I think this may be the same as #3300 after all.

I think the problematic code is this: https://github.com/bundler/bundler/blob/master/lib/bundler/definition.rb#L205

```
sources.all_sources.each do |s|
  s.dependency_names = dependency_names.dup
  idx.add_source s.specs
  s.specs.each { |spec| dependency_names.delete(spec.name) }
  dependency_names.push(*s.unmet_deps).uniq!
end
```

`sources`, when the block syntax is used, includes at least two Source objects, each with a single 'remote'.

`dependency_names` starts out as the full list of gems from the Gemfile, but gets whittled down source-by-source. The intention is for Gem dependencies to be distributed across sources, but in my extensive haphazard `puts` debugging, `s.specs` for the first source contains gems from all sources. My guess is that it includes info about locally cached gems, not just remote ones.

The effect seems to be that `bundle outdated` will only ever tell you about updates for the first source to appear in `sources.all_sources`.

I'm not sure what a valid fix is; seems like the conditions that trigger `dependency_names.delete` need to be more strict. Ideally, something would only get deleted from `dependency_names` when it was determined that a source truly 'owned' it, but for Rubygems sources, `s.specs` seems to just mash together the remote specs + all locally installed specs, making it hard to audit where they came from afterward.

**Comment by TimMoore** [comment]
This should be fixed in 1.7.13. Please try upgrading and let us know if you still have any problems.
